File: atcoder/past17-2/h.py
from collections import defaultdict
import itertools

INF=100000000000000
n,m=map(int,input().split())
classes=[list(map(int,input().split())) for _ in range(n)]
#dp[単位]=必要な労力
dp=defaultdict(lambda: INF)
dp[0]=0
for _,l in itertools.groupby(classes,lambda x: x[1]):
    l = list(l)
    for k in list(dp.keys()):
        for a,b,c in l:
            nk = min(k+c,m)
            dp[nk] = min(dp[nk],dp[k]+a)

if dp[m]==INF:
    print(-1)
else:
    print(dp[m])

File: atcoder/past17-2/h.rb
INF=100000000000000
n,m=gets.split.map(&:to_i)
classes=(1..n).map{gets.split.map(&:to_i)}
#dp[単位]=必要な労力
dp=Hash.new{INF}
dp[0]=0
classes.group_by{|a,b,c|b}.each{|_,l|
    dp.keys.each{|k|
        l.each{|a,b,c|
            nk=[k+c,m].min
            dp[nk]=[dp[nk],dp[k]+a].min
        }
    }
}
p dp[m]==INF ? -1 : dp[m]